### ISSUE TITLE
error when run

### ISSUE THREAD

**Issue by OverDriveMC** [horovod/horovod#402]
when I run deepFM using horovod,I meet this error:

> INFO:tensorflow:An error was raised. This may be due to a preemption in a connected worker or parameter server. The current session will be closed and a new session will be created. Error: Horovod has been shut down. This was caused by an exception on one of the ranks or an attempt to allreduce, allgather or broadcast a tensor after one of the ranks finished execution. If the shutdown was caused by an exception, you should see the exception in the log before the first shutdown message.
	 [[Node: DistributedAdamOptimizer_Allreduce/HorovodAllgather_gradients_concat_12_0 = HorovodAllgather[T=DT_FLOAT, _device="/job:localhost/replica:0/task:0/device:CPU:0"](gradients/concat_12/_14449)]]
	 [[Node: DistributedAdamOptimizer_Allreduce/HorovodAllgather_gradients_dnn_input_layer_1_itemCity_embedding_itemCity_embedding_weights_embedding_lookup_sparse_embedding_lookup_grad_Reshape_1_0/_11881 = _HostRecv[client_terminated=false, recv_device="/job:localhost/replica:0/task:0/device:GPU:0", send_device="/job:localhost/replica:0/task:0/device:CPU:0", send_device_incarnation=1, tensor_name="edge_6023_...eshape_1_0", tensor_type=DT_INT64, _device="/job:localhost/replica:0/task:0/device:GPU:0"]()]]

> Caused by op u'DistributedAdamOptimizer_Allreduce/HorovodAllgather_gradients_concat_12_0', defined at:
  File "single_train_model_fmnn_video_mgpu.py", line 166, in <module>
    tf.app.run()
  File "/home/aronrwang/env/ringgpu/Python27/lib/python2.7/site-packages/tensorflow/python/platform/app.py", line 126, in run
    _sys.exit(main(argv))
  File "single_train_model_fmnn_video_mgpu.py", line 150, in main
    m.train(input_fn=lambda: input_fun(sfile[hvd.local_rank()], num_epochs=2),hooks=[bcast_hook]) ##change for train time:same with baseline
  File "/home/aronrwang/env/ringgpu/Python27/lib/python2.7/site-packages/tensorflow/python/estimator/estimator.py", line 352, in train
    loss = self._train_model(input_fn, hooks, saving_listeners)
  File "/home/aronrwang/env/ringgpu/Python27/lib/python2.7/site-packages/tensorflow/python/estimator/estimator.py", line 812, in _train_model
    features, labels, model_fn_lib.ModeKeys.TRAIN, self.config)
  File "/home/aronrwang/env/ringgpu/Python27/lib/python2.7/site-packages/tensorflow/python/estimator/estimator.py", line 793, in _call_model_fn
    model_fn_results = self._model_fn(features=features, **kwargs)
  File "/data1/home/aronrwang/chengmeng/to_pub/net/fmcdnn.py", line 479, in _model_fn
    config=config)
  File "/data1/home/aronrwang/chengmeng/to_pub/net/fmcdnn.py", line 416, in _deepfm_model_fn
    logits=logits)
  File "/home/aronrwang/env/ringgpu/Python27/lib/python2.7/site-packages/tensorflow/python/estimator/canned/head.py", line 1153, in create_estimator_spec
    train_op=train_op_fn(regularized_training_loss))
  File "/data1/home/aronrwang/chengmeng/to_pub/net/fmcdnn.py", line 390, in _train_op_fn
    scope=dnn_parent_scope)))
  File "/home/aronrwang/env/ringgpu/Python27/lib/python2.7/site-packages/tensorflow/python/training/optimizer.py", line 359, in minimize
    grad_loss=grad_loss)
  File "/home/aronrwang/env/ringgpu/Python27/lib/python2.7/site-packages/horovod/tensorflow/__init__.py", line 185, in compute_gradients
    device_sparse=self._device_sparse)
  File "/home/aronrwang/env/ringgpu/Python27/lib/python2.7/site-packages/horovod/tensorflow/__init__.py", line 71, in allreduce
    values = allgather(tensor.values)
  File "/home/aronrwang/env/ringgpu/Python27/lib/python2.7/site-packages/horovod/tensorflow/mpi_ops.py", line 111, in allgather
    return MPI_LIB.horovod_allgather(tensor, name=name)
  File "<string>", line 65, in horovod_allgather
  File "/home/aronrwang/env/ringgpu/Python27/lib/python2.7/site-packages/tensorflow/python/framework/op_def_library.py", line 328, in apply_op
    op_type_name, name, **keywords)
  File "/home/aronrwang/env/ringgpu/Python27/lib/python2.7/site-packages/tensorflow/python/framework/op_def_library.py", line 787, in _apply_op_helper
    op_def=op_def)
  File "/home/aronrwang/env/ringgpu/Python27/lib/python2.7/site-packages/tensorflow/python/framework/ops.py", line 3271, in create_op
    op_def=op_def)
  File "/home/aronrwang/env/ringgpu/Python27/lib/python2.7/site-packages/tensorflow/python/framework/ops.py", line 1650, in __init__
    self._traceback = self._graph._extract_stack()  # pylint: disable=protected-access

I use 3 gpu, I divide the data  into three parts, but rank 0 gpu have more data than others, I see the other two gpu can finish successfully, I doubt whether it is caused of the unequal distribution data?
I set the epoch=2,  because the other 2 gpu have less data, they break short, and the rank 0 gpu can't receive any data, so I cause this error?
does this mean I must divide  the data  into equal  amounts?

**Comment by TimYuenior** [comment]
Is it possible to try it with equal amounts of data? As far as I know, Horovod does synchronous updates, so I would imagine that the rank 0 GPU is waiting for data which it doesn't get since the other GPUs are finished. I believe that's what this line means:
> This was caused by an exception on one of the ranks or an attempt to allreduce, allgather or broadcast a tensor after one of the ranks finished execution.

Might be wrong though, I'm no expert.

**Comment by OverDriveMC** [comment]
@TimYuenior  The data consists of 200 parts of file, I run the model with 4 GPU or 2 GPU, It can run successfully, So I guess this is caused of unequal distribution data.
